### PR TITLE
[3.6] UPSTREAM: 58572: Automated cherry pick of #58547: Send correct resource version for delete events from watch

### PR DESCRIPTION
**Pull request by tnozicka**
Backport of #18233 via kubernetes/kubernetes#58572

/assign @liggitt

### CODE REVIEW

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher.go [hunk 1]
```diff
@@ -880,6 +882,15 @@ func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
...
+		// return a delete event with the previous object content, but with the event's resource version
+		object, err := c.copier.Copy(event.PrevObject)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("unexpected copy error: %v", err))
+			return
+		}
+		if err := c.versioner.UpdateObject(object, event.ResourceVersion); err != nil {
+			utilruntime.HandleError(fmt.Errorf("failure to version api object (%d) %#v: %v", event.ResourceVersion, object, err))
+		}
 		watchEvent = watch.Event{Type: watch.Deleted, Object: object}
```
Comment (liggitt) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: this looks like it was completely wrong before, not sending PrevObject in delete events...

Comment (tnozicka) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: https://github.com/kubernetes/kubernetes/commit/e9e69356e4907fa4d0f45ea7e7768357ba71aba9

Comment (liggitt) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: move `object, err := c.copier.Copy(event.Object)` into the added/modified cases then?

Comment (liggitt) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: and update the description to indicate this is a pick of those two fixes

**Comment by tnozicka** [comment]
@liggitt picked both commit properly; PTAL

**Comment by tnozicka** [comment]
no chance to pass on borked CI which won't be fixed https://github.com/openshift/origin/issues/18276#issuecomment-360283477

**Pull request closed without merging**